src/Arriba/Arriba/Model/Table.cs

```diff
@@ -513,6 +506,18 @@ public void AddOrUpdate(DataBlock.ReadOnlyDataBlock values, AddOrUpdateOptions o
             }
         }
 
+        private void VerifyNoColumnsAreMissing(DataBlock.ReadOnlyDataBlock values)
+        {
+            foreach (ColumnDetails column in values.Columns)
+            {
+                ColumnDetails foundColumn;
+                if (!_partitions[0].DetailsByColumn.TryGetValue(column.Name, out foundColumn))
+                {
+                    throw new ArribaException(StringExtensions.Format("AddOrUpdate failed because values were passed for column '{0}', which is not in the table. Use AddColumn to add all columns first or ensure the first block added to the Table has all desired columns.", column.Name));
```

Comment:
There is a specific exception for ColumnNotFoundException which could be more appropriate here.

Reply:
+1. This was a refactoring so didn't want to change the behavior